server/container/core/src/main/java/org/apache/james/core/ReferenceCount.java

```diff
@@ -0,0 +1,75 @@
+/****************************************************************
+ * Licensed to the Apache Software Foundation (ASF) under one   *
+ * or more contributor license agreements.  See the NOTICE file *
+ * distributed with this work for additional information        *
+ * regarding copyright ownership.  The ASF licenses this file   *
+ * to you under the Apache License, Version 2.0 (the            *
+ * "License"); you may not use this file except in compliance   *
+ * with the License.  You may obtain a copy of the License at   *
+ *                                                              *
+ *   http://www.apache.org/licenses/LICENSE-2.0                 *
+ *                                                              *
+ * Unless required by applicable law or agreed to in writing,   *
+ * software distributed under the License is distributed on an  *
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY       *
+ * KIND, either express or implied.  See the License for the    *
+ * specific language governing permissions and limitations      *
+ * under the License.                                           *
+ ****************************************************************/
+
+package org.apache.james.core;
+
+import org.apache.james.lifecycle.api.LifecycleUtil;
+
+import com.google.common.base.Preconditions;
+
+public class ReferenceCount<Id, ExceptionId extends Exception> {
+
+    public interface CloningOperation<Id, ExceptionId extends Exception> {
+        Id clone(Id clonee) throws ExceptionId;
+    }
+
+    private Id reference;
```

Comment:
No. We have a reference = null used below as a gc optimisation technic, that forbids us to be final.